File: Graph_Algorithms/DFS.c
#include <stdio.h>
#include <stdlib.h>

struct Stack {

	int* arr;
	int top;
	int size;
};

struct Stack* create (int size) 
{
	struct Stack* stack = NULL;

	stack = (struct Stack*)malloc(sizeof(struct Stack));
	stack->top = -1;
	stack->arr = (int*)malloc(sizeof(int)*size);	
	stack->size = size - 1;

	return stack;
}

int empty (struct Stack* stack)
{

	return (stack->top == -1);
}

int full (struct Stack* stack)
{

	return (stack->top > stack-> size);
}

void push (struct Stack* stack,int x)
{
	if (!full(stack)) {
		
		stack->top = stack->top + 1;
		stack->arr[stack->top] = x;
		
	}
}

int pop(struct Stack* stack)
{
	if (!empty(stack)) {

		int x;
	
		x = stack->arr[stack->top];
		stack->top  = stack->top - 1;
		
		return x;
	}	
}

void dfs (int s,int g[][1000],int state[],struct Stack* stack,int v)
{
	int i;
	int x;

	push(stack,s);										//Push Source node in Stack
	printf("DFS \n");
	
	while (!empty(stack)) {

		x = pop(stack);

		if (state[x] ==  -1) {

			printf("%d ",x);
			state[x] = 1;
		}

		for (i = 0; i < v; i++) {
	
			if ((g[x][i] == 1) && (state[i] == -1)) {

				push(stack,i);
			
			}
	
		}
			
	}
	printf("\n");
}

int main()
{
	int max;
	int v;
	int e;
	int i;
	int j;
	int x;
	int y;
	int s;
	int g[1000][1000];    							//Adjancency Matrix 'g'of int of  1000x1000

	scanf("%d %d",&v,&e);

	for (i = 0; i < v; i++) {

		for (j = 0; j < v ; j++) {

			g[i][j] = 0;
		}
	}

	
	for (i = 0; i < e; i++) {

		scanf("%d %d",&x,&y);						//Nodes with Edges in Undirected Graph

		g[x][y] = 1;
		g[y][x] = 1;
	}

	scanf("%d",&s);								//Take 's' as a source node

	int state[v];

	for (i = 0; i < v; i++) {
	
		state[i] = -1;							//All Nodes are Unvisited initially
	}

	struct Stack* stack = NULL;
	stack = create(v);
	
	dfs(s,g,state,stack,v);

	return 0;
}

File: Graph_Algorithms/dfs.py
stack = []

def empty() :
	if len(stack) == 0:
		return 1
	else :
		return 0

def push (x) :
	if len(stack) > n-1 :
		#print("overflow") 
		return 666
	else :
		stack.append(x)
		#print(stack)	
	
def pop() :
	if len(stack) == 0:
		#print("underflow")
		return 666
	else :
		x = stack.pop()
		#print("after pop")
		#print(stack)
		return x


g = []
vis = []

print('Enter no. of nodes')		#No. of nodes in graph

n = input()
n = (int(n))

print('Enter no. of edges')
e = input()
e = (int(e))

for i in range (n) :
	g.append([])
	for j in range (n) :
		g[i].append(0)

print('Enter the Connection between edges')   #Enter the Edges in graph

for i in range (e) :
	
		x = input()
		x = ( int(x))
		y = input()
		y = (int(y))
		g[x][y] = 1
		g[y][x] = 1

for i in range (n) :
	vis.append(-1)

print('Enter source node')		#Enter Source node

s = input()
s = (int(s))

vis[s] = 0

push(s)							#Push Source in queue

def dfs (g,stack,vis,s) :

		while(not(empty())) :
			x = pop()
			
			if (vis[x] == -1):
				print(x)
				vis[x] = 1

			for i in range (n) :
				if (g[x][i] == 1):
					if(vis[i] == -1) :
						push(i)
			

dfs(g,stack,vis,s)


			




